### ISSUE TITLE
Enhancement: gns3_server vs gns3_controller and Linux vs Windows

### ISSUE THREAD

**Issue by josephmhiggins** [GNS3/gns3-server#2062]
With
https://github.com/GNS3/gns3-server/issues/2059, and
https://github.com/GNS3/gns3-server/issues/2060, and
https://github.com/GNS3/gns3-server/issues/2061

I can see their is major architectural flaw in GNS3.
There is no gns3_controller.log
I have no such thing or any book that leads me to logging an abstract entity as the gns3_controller.
With the caveat that you can not take anyone verbatim,
Jeremy Grossman and Julien have loosely publicly stated that the "gn3 controller" knows everything.

The root core problem is that Linux users can do something like
gns3 --version, or gns3 --help, or gns3 --debug 2 and I theoretically can do in windows something like this in powershell:
ProcessStartInfo Class
p.StartInfo = "--debug """2"""

But there is nothing there in Windows.
Linux users have all these fancy parameters that they can pass to gns3
I can not pass anything.

With Hyper-V, Windows looks to be completely legitimate as a host/server.

So in Windows 10 I can create a wrapper around gns3 and use start  process info
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo?view=net-6.0
(I can do and have done and have backed up all that stuff now)

to do
start-gns3-remote-controller.ps1 vbox
and in gns3-remote-controller.ps1
I can ......*thinking*
if ($arg -eq vbox) {
    ...thinking..start vbox controller log
}

The absolute point of demarcation is ME.
Now....for the gns3 developers, the point of demarcation is the gns3 controller because they can NOT introduce a layer.
**I will handle building a gns3 controller layer in powershell.**

But I want all parameters that Linux Users have to start gns3 available to me as a Windows user.

GNS3 developers call the thing that starts GNS3 Profiles.
It is NOT profiles.
Profiles is a subargument.
It is "GNS3 Controller - Profile".
I will handle building - for me only - the "GNS3 Controller - Profile" wrapper in powershell.
I will handle building - for me only the gns3_controller.log in powershell.

The point of demarcation of gns3 is not the gns3 controller - it is the gns3 user.
I will handle building the point of demarcation - for me only.

I need all all parameters that Linux Users have available, available to me as a Windows user - I want to pass the profile to 
GNS3.

Because the gns3 developers have decided to open up the gns3_server.log file in write mode, instead of append mode, I have
to fix all my logs and piece them together. 

If I had a powershell wrapper built and and the gns3 program in Windows accepting parameters like Linux does, then I would not have to spend hours piecing together all my gns3_server.log files.

So, new GNS3 users can not handle any of this.
But this is not fair that Linux can pass parameters and I can not pass parameters in Windows to GNS3.

**Comment by josephmhiggins** [comment]
Oh, I want to paste a working powershell example of my code, which is all around #2059, #2060, and #2061.
I am not going to make it pretty. The code is what it is:

# C:\Users\joe_I7\OneDrive\Documents\software_versioning\powershell\wireshark\InitWireshark.ps1
# https://stackoverflow.com/questions/371816/how-to-clear-the-wireshark-filter-combo-contents
# process explorer
# "C:\Program Files\GNS3\tail.exe" -f -c +0b "C:\Users\joe_I7\GNS3\projects\docker-cloud-export\project-files\captures\gns3-endhost-1_eth0_to_Cloud1_eth2.pcap"
$MyCaptureFile = $HOME + "\_justsomefile.pcapng"
$MyLogfile = $HOME + "\_wireshark.log"
$MyInterfaceName = "SW241-G6"
$MyLibpcapFilter = "icmp"
# version 0 $MyWindowTitle = "gui.window_title: External Network ICMP"#  + $MyLibpcapFilter.ToUpper()
# works#1$MyWindowTitle = "gui.window_title:External Network:ICMP only"
$MyWindowTitle = '"gui.window_title:External Network:ICMP only"' # works #2
$pinfo = New-Object System.Diagnostics.ProcessStartInfo
$pinfo.FileName = 'C:\Program Files\Wireshark\wireshark.exe'
$InterfaceName = "SW241-G6"
$MyLibpcapFilter = "icmp"

# this prints the whole error
# $pinfo.RedirectStandardError = $true
# $pinfo.UseShellExecute = $false
# OnExisted$pinfo.
$pinfo.Arguments = @(
    "-k"
    # works#1 "-o ""gui.window_title:External Network:ICMP only"""
    "-o $MyWindowTitle"    # works #2
    "-i $MyInterfaceName"
    "-f $MyLibpcapFilter"
    "-w $MyCaptureFile"
    "--log-file $MyLogFile"
)
$p = New-Object System.Diagnostics.Process
$p.StartInfo = $pinfo
$p.Start() | Out-Null
# this prints the whole error
# Start-Sleep 2
# $stderr = $p.StandardError.ReadToEnd()
# $stderr | Out-File $MyLogfile
<#
if ($p.ExitCode -ne 0 ) {
    Write-Host "Found undefined exit code:" $p.ExitCode
    # Write-Host "stderr: $stderr" -ForegroundColor White
    return
} else {
    Write-Host No Errors
}
#>
# $p.WaitForExit()

**Comment by eantowne** [comment]
![gns3-windows-cmd-line](https://user-images.githubusercontent.com/66289315/165934593-16fc5bf7-8133-49c7-ab7e-ff3344e97610.png)
![gns3server-windows-options](https://user-images.githubusercontent.com/66289315/165934610-aea93baa-c93e-473b-bd73-9a2cfc9c655b.png)

It appears to me that the Windows binaries have all of the same options as the Linux versions. Or are you saying that they do not work? If they do not work as expected, then that should be a separate issue.

Controller.log should not be a thing (from my perspective) as the controller is simply part of the gns3server, and thus covered under it's log. If this is not acceptable, it should also be a separately submitted issue.

As for the controller issue, my understanding is that the controller is run by the 1st server the gui connects to, or simply the 1st or "primary" server. So, if you start the Windows GUI for the first time or run the wizard, you can choose to use "VM" or "remote" as the primary server, thus the controller will be on that server. My understanding is that you want to completely separate the controller as a new binary from the gns3server binary? What benefit does that provide, when you can just use the Windows "local" server as the primary, and thus have it be the controller. Once that is done you can simply connect to local GNS3 VM's as remote, and remotes as remote. Although, I am not entirely sure what the benefit of having the controller running on the local server in Windows actually provides you.

**Comment by josephmhiggins** [comment]
Thanks Ean on the .\gns3 --help
It works as you stated.
What I stated "I can not pass anything." was then completely wrong.

Actually, I did not want to bring your name into it, but I want to do the things that you do.
So if I run the local server option with a virtualbox/gns3 vm or a vmware/gns3 vm or a hyper-v/gns3 vm, then
it is still the same controller and the log gets appended to the same logfile. I will handle implementing that.

Now, today, I have 3 differently named gns3_server.log files because the gns3 developers open up the gns3-server.log
in write mode. I already implemented a wrapper around gns3 so that I kept all my gns3_server.log files based on
a start/stop time and that succeeded. However, I did not keep pace with the gns3 developers when I implemented
profiles. My wrapper fell apart. Troubleshooting #2061 (that should be the main culprit) is a major pain because
I now have over 30 tabs open and some tabs have 20 sub-tabs open. It's too confusing.

Based on your information, I can see that I have --profile "available" to me. Based on your information, then I think I can work with it and build a new wrapper.

"My understanding is that you want to completely separate the controller as a new binary from the gns3server binary?"
No, I do not want to do that.

I think you have totally solved all my concerns - except for the code that I will write and test.

I can start gns3 as a local server or remote server and that will be up to me to figure out how to log things with powershell.

**Comment by eantowne** [comment]
"open up the gns3-server.log in write mode" may be an option for a separate issue. From my perspective, logs should never be opened in write mode, and should always be written to as an append.

**Comment by josephmhiggins** [comment]
I agree on the append mode for logs...but it is a whole big thing because I can change something in the gns3_gui.ini and affects my controller/profile/gns3 project that is something I will have to deal with on my own. I do not think the gns3 developers should be getting involved in how I manage my gns3.

**Comment by eantowne** [comment]
I am looking at the code, it looks like the devs are using the default logging library in python. I am trying to determine if there is a way to change the file handler to use an append mode vs write mode.

**Comment by eantowne** [comment]
According to the python documentation, the logging default configuration is to use append. Having a hard time trying to find where they are overwriting it. Going to clone the repo so I can search better.

**Comment by eantowne** [comment]
So, they are using RotatingFileHandler(). This uses append by default and I am not seeing that they are overriding the mode, so not sure what is going on.

Actually, this may be the issue https://github.com/GNS3/gns3-server/blob/99813b3cfa15498f080877cc0920053c380814ef/gns3server/web/logger.py#L137

Just to verify, you are not running any scripts anywhere that are accessing the log file(s) simultaneously?

I would recommend opening a separate issue for the dev team specific to this issue.

**Comment by josephmhiggins** [comment]
Just to verify, you are not running any scripts anywhere that are accessing the log file(s) simultaneously?

No - "I" am not running any scripts that I wrote that accessing the log file simultaneously.
However, on 20220426_1048UTC I introduced into .gvimrc file

" To make Vim automatically refresh any files that haven't been edited by Vim
set autoread

That change was made to troubleshoot #2061
I made it so I can read the packets dropped in real time by the windows firewall.

The autoread option, to the best of my knowledge, works like the
"watch" command linux. Some files I can put a lock on - I do not know how to do it but I am pretty sure it can be done - so that nothing else can read the file. The gns3 developers do not do anything like that. So, the gns3_server.log is updated in real time when I use gvim. If I do not have autoread enabled, then when I move out of gvim - as the ActiveWindow - and move into another windows - a new ActiveWindow - and then move back to gvim and reactivate gvim as the ActiveWindow, then gvim issues a popup:
"gns3_server.log has changed would you like to reload the file: Load File    Cancel"

**Comment by eantowne** [comment]
Linux troubleshooting:

1. Started gns3server (by itself) using the following command:
`gns3server --config /home/ean/.config/GNS3/2.2/gns3_server.conf --log ~/.config/GNS3/2.2/gns3_server_stand_alone.log`
2. Was able to `tail -f` the server log
3. Started GNS3 GUI, client connect log entry visible in `tail -f`
4. Opened the log file in VS Codium, added "test" on a new line, saved the file
5. `tail -f` displayed the change

Summary:
On Linux I am not seeing an issue with the file being locked out by the gns3server when logging.

Next, testing on Windows (11)

**Comment by eantowne** [comment]
@josephmhiggins Can you try a similar testing methodology on a clean boot with Windows? I have basically nothing installed on my 1 small Windows system.

**Comment by josephmhiggins** [comment]
On Windows, you always want to always try to get Windows users to use Notepad.
I right click on the gns3_server.log - open with notepad
type in text
Save file
close the file
open the file
It's there:
2022-03-26 06:23:19 INFO run.py:272 Remove PID file C:\Users\joe_I7\AppData\Roaming\GNS3\2.2\gns3_server.pid
test

A clean boot is for when everything is totally destroyed and nothing makes sense, so that is out of the question.

I would use ProcessExplorer to see if a file is locked - it's in one of my books and I already posted it on gns3.com how to do it; albeit processexplorer is very complicated to use.
or
I would use one of my powershell scripts. It's a 'try' 'catch' to see if a file is locked - it's kind of easy and the try catch is searchable right away on the web; it's standard powershell stuff.

The important thing is can I read the file and I know I can read the file.

So I can read the file and copy it's updates realtime somehow - I do not know how to do that - into another file and then do something like:
2022-03-26 06:23:19 INFO run.py:272 Remove PID file C:\Users\joe_I7\AppData\Roaming\GNS3\2.2\gns3_server.pid
2022-03-26 06:23:20 HEYYOU USER joe_I7 just changed GNS3 Server Host binding to 1.1.1.1

Actually that would be from monitoring gns3_gui.ini also, but it's the same idea.

A very tasty one would be copying from gns3_server.log
2022-03-26 06:23:20 HEYYOU USER joe_I7 closing oldlab.gns3 and opening up newlab.gns3

and appending to $HOME\gns3higgins\gns3_somecontroller.log
e.g. gns3_localcontroller.log

Hrmm. I like that.

**Comment by josephmhiggins** [comment]
Function Test-FileLock {
    [CmdletBinding()]
    param (
        [parameter(Mandatory=$true)][string]$Path
    )

  $oFile = New-Object System.IO.FileInfo $Path

  if ((Test-Path -Path $Path) -eq $false) {
    return $false
  }

  try {
    $oStream = $oFile.Open([System.IO.FileMode]::Open, [System.IO.FileAccess]::ReadWrite, [System.IO.FileShare]::None)

    if ($oStream) {
        $oStream.Close()
    }
      $false
    } catch {
        # file is locked by a process.
        return $true
    }
}

**Comment by josephmhiggins** [comment]
Per Ean's information. This is resolved.
